tests/rbd/test_performance_immutable_cache.py

```diff
@@ -0,0 +1,249 @@
+        str: Time taken for the I/O operation.
+    """
+    config = kw.get("config")
+    rbd = rbd_obj.get("rbd")
+
+    # Write Data to the Parent Image
+    fio_args = {
+        "client_node": client,
+        "pool_name": pool,
+        "image_name": image,
+        "io_type": "write",
+        "size": config["fio"]["size"],
+    }
+
+    run_fio(**fio_args)
+
+    # Create snapshot, protect, and clone for the image
+    snap = "image_snap" + random_string(len=5)
+    clone = "image_clone" + random_string(len=5)
+    snap_spec = f"{pool}/{image}@{snap}"
+    clone_spec = f"{pool}/{clone}"
+
+    create_snap_and_clone(rbd, snap_spec, clone_spec)
+
+    # After image clone performing read operation from parent and clone
+
+    fio_args = {
+        "client_node": client,
+        "pool_name": pool,
+        "image_name": image,
+        "io_type": "read",
+        "size": config["fio"]["size"],
+        "get_time_taken": True,
+    }
+
+    config["time_taken"] = run_fio(**fio_args)
+    time = config.get("time_taken", "")
+
+    fio_args = {
+        "client_node": client,
+        "pool_name": pool,
+        "image_name": clone,
+        "io_type": "read",
+        "size": config["fio"]["size"],
+        "get_time_taken": True,
+    }
+
+    config["time_taken"] = run_fio(**fio_args)
+    time += config.get("time_taken", "")
+
+    return time
+
+
+def test_immutable_cache_performance(rbd_obj, **kw):
+    """
+    Test the performance of Immutable Object Cache with and without cache.
+
+    Args:
+        rbd_obj (object): RBD object.
+        kw (dict): Test data.
+
+    Raises:
+        Exception: If any test step fails.
+    """
+    ceph_cluster = kw.get("ceph_cluster")
+    client = ceph_cluster.get_nodes(role="client")[0]
+
+    try:
+        # Configure Immutable Object Cache
+        if configure_immutable_cache(rbd_obj, client, "user1"):
+            log.error("Immutable object cache configuration failed")
+            return 1
+        log.info("Immutable object cache configuration completed successfully.")
+
+        for pool_type in rbd_obj.get("pool_types"):
+            rbd_config = kw.get("config", {}).get(pool_type, {})
+            multi_pool_config = getdict(rbd_config)
+
+            for pool, pool_config in multi_pool_config.items():
+                multi_image_config = getdict(pool_config)
+
+                for image in multi_image_config.keys():
+                    # Perform I/O operations with cache enabled
+                    time_with_cache = perform_clone_io(
+                        client=client,
+                        rbd_obj=rbd_obj,
+                        pool=pool,
+                        image=image,
+                        io_type="write",
+                        **kw,
+                    )
+
+                    # Removing immutable cache
+                    client.exec_command(cmd=f"rm -rf {immutable_cache_path}", sudo=True)
+
+                    image = image + random_string(len=3)
```

Comment:
Are you creating a different image without cache enabled and testing the performance? I feel it will be better if you just first test on the same image without cache enabled, then enable cache on the same image and then compare both time taken.

Reply:
Done, Tried it on same image.

Reply:
when i tried it on the same image time difference is very less may be due to some existing OS cache,
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-APY3PE/verify_performance_for_Immutable_object_cache_0.log
but when i tried with new image time difference is more and test are passing always
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-VKNWDX

Reply:
Yea, but Ilya did mention that the time difference may not be properly conceivable in his chat. Could you check with him once if its okay to test with different images?

Reply:
ok i tried first without cache and next with cache on the same image it worked,
Thanks for the suggestion.